[SaferCPP] Use smart pointers in LinkIconCollector

zakariaridouh opened the pull request:
#### 85b73124f671d6d8fbe2b6dc730b98711b42326d
<pre>
[SaferCPP] Use smart pointers in LinkIconCollector
<a href="https://bugs.webkit.org/show_bug.cgi?id=291134">https://bugs.webkit.org/show_bug.cgi?id=291134</a>
&lt;<a href="https://rdar.apple.com/problem/148648198">rdar://problem/148648198</a>&gt;

Reviewed by Chris Dumez.

Use smart pointers/references in LinkIconCollector.h
and LinkIconCollector.cpp

* Source/WebCore/html/LinkIconCollector.cpp:
(WebCore::LinkIconCollector::iconsOfTypes):
* Source/WebCore/html/LinkIconCollector.h:
(WebCore::LinkIconCollector::LinkIconCollector):

Canonical link: <a href="https://commits.webkit.org/293327@main">https://commits.webkit.org/293327@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/fc4453b7a5fd071d9c61d30984523a981b2fb3b2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/98582 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/18215 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/8451 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/103703 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/49140 "Built successfully") 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/100626 "Build is in progress. Recent messages:OS: Sequoia (15.3), Xcode: 16.2; Skipping applying patch since patch_id isn't provided; Checked out pull request; Passed bindings tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/18508 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/26666 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/75040 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/32196 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/101586 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/132/builds/14039 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/89030 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/55398 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/13818 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/6995 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/48552 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/83774 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/7073 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/106076 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/25672 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/18697 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/84017 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/26049 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/85231 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/83502 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/21094 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/28142 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/5821 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/19355 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/25630 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/30811 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/25448 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/28768 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/27023 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/html/LinkIconCollector.h:
```diff
@@ -35,15 +35,15 @@ enum class LinkIconType : uint8_t;
 
 private:
-    Document& m_document;
+    Ref<Document> m_document;
```
Likely should be marked as `const`

cdumez commented on Source/WebCore/html/LinkIconCollector.cpp:
```diff
@@ -35,7 +35,7 @@
 namespace WebCore {
 
-const unsigned defaultTouchIconWidth = 60;
+static constexpr unsigned defaultTouchIconWidth = 60;
```
You don't need static, `const` or `constexpr` at namespace level implies `static`

zakariaridouh replied:
Got it, so just `unsigned defaultTouchIconWidth = 60;` is fine?

cdumez replied:
`constexpr unsigned defaultTouchIconWidth = 60;`

cdumez replied:
Making `constexpr` is good. I was merely objecting about adding `static` since it is redundant here.

zakariaridouh replied:
Makes sense, I misread the sentence.

cdumez commented on Source/WebCore/html/LinkIconCollector.cpp:
```diff
@@ -75,43 +75,43 @@ static int compareIcons(const LinkIcon& a, const LinkIcon& b)
     }
 
-    std::sort(icons.begin(), icons.end(), [](auto& a, auto& b) {
+    std::sort(icons.begin(), icons.end(), [](const auto& a, const auto& b) {
```
We don't usually use `const` with `auto`

cdumez commented on Source/WebCore/html/LinkIconCollector.h:
```diff
@@ -35,15 +35,15 @@ enum class LinkIconType : uint8_t;
 class LinkIconCollector {
 public:
-    explicit LinkIconCollector(Document& document)
+    explicit LinkIconCollector(Ref<Document> document)
```
We do not pass smart pointers by value due to ref-counting churn. Please use a r-value reference and `WTFMove()` below.

cdumez commented on Source/WebCore/html/LinkIconCollector.cpp:
```diff
@@ -35,7 +35,7 @@
 namespace WebCore {
 
-const unsigned defaultTouchIconWidth = 60;
+unsigned defaultTouchIconWidth = 60;
```
Please add back the `constexpr` here.

cdumez commented on Source/WebCore/html/LinkIconCollector.cpp:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (C) 2016 Apple Inc. All rights reserved.
+ * Copyright (C) 2025 Apple Inc. All rights reserved.
```
2016-2025

cdumez commented on Source/WebCore/html/LinkIconCollector.h:
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (C) 2016 Apple Inc. All rights reserved.
+ * Copyright (C) 2025 Apple Inc. All rights reserved.
```
2016-2025

cdumez reviewed (approved):
LGTM with fixes.

zakariaridouh committed:
[SaferCPP] Use smart pointers in LinkIconCollector
https://bugs.webkit.org/show_bug.cgi?id=291134
<rdar://problem/148648198>

Reviewed by Chris Dumez.

Use smart pointers/references in LinkIconCollector.h
and LinkIconCollector.cpp

* Source/WebCore/html/LinkIconCollector.cpp:
(WebCore::LinkIconCollector::iconsOfTypes):
* Source/WebCore/html/LinkIconCollector.h:
(WebCore::LinkIconCollector::LinkIconCollector):

Canonical link: https://commits.webkit.org/293327@main
